### PR TITLE
Calculate set difference to avoid unnecessary copy to S3

### LINKED ISSUE #32: Efficiency concerns
This was prompted by https://github.com/thoth-station/thoth-application/issues/2604

We currently spawn two subprocess per document to sync, if I read app.py
correctly. Could we not use `aws sync s3://src s3://dest` ? Or is that too
coarse grained ?

### LINKED ISSUE #2604: [spike][3pt]document is not available for few packages
**Describe the bug**
Document-sync job is not sycning all the documents.

**Acceptance criteria**
- [ ]  check if the [document sync job](https://github.com/thoth-station/thoth-application/blob/master/document-sync-job/overlays/ocp4-stage/cronjob.yaml) is still in progress.  
- [ ] research on what documents are missing. 

Additional notes:
 - the document sync might be still in progress

### PR DESCRIPTION
**Pull request by VannTen**
Intent: We list object from the source and the destination and calculate the set difference.
Both query use the list_objects_v2 from boto3 since we're talking to s3 store.
The keys are guaranteed (by the S3 api def) to be sorted, so we can do the set
difference lazily by walking each generator (see `sorted_set_diff`).
-> This is to avoid checking objects existence in the destination one by one

Also, we use boto3 directly for upload and download which will avoid the
fork-exec and the need for a temporary file.

NOTE TO REVIEWERS: 
I've arranged the commits into logical chunks, so they should be mostly
individuably reviewable.

EDIT:
fix #32 
fix thoth-station/thoth-application#2604

### CODE REVIEW

**Comment by VannTen** [comment]
Intend to fix #32

**Comment by VannTen** [comment]
Should be ready for review o/

@harshad16 : do we have a test suite / test cluster to check the functionnality
? From what I've seen while coding, we probably had other problems than
excessive memory or timeout (-> related to storages and handling solver
prefixes)

**Comment by VannTen** [comment]
Also, this might be a candidate for upstreaming (into boto3 / botocore maybe).
There is a feature request https://github.com/boto/boto3/issues/358 which is open for some time.
@goern

**Comment by goern** [comment]
> Also, this might be a candidate for upstreaming (into boto3 / botocore maybe). There is a feature request [boto/boto3#358](https://github.com/boto/boto3/issues/358) which is open for some time. @goern

feel free to contribute back to upstream, if we can do it in a reasonable amount of time...

**Comment by VannTen** [comment]
I'll go probe them and see how that goes.

**Comment by harshad16** [comment]
> Should be ready for review o/

Can you please rebase your commit to 1? as there only one feature we are working on here.


> do we have a test suite / test cluster to check the functionnality ? From what I've seen while coding, we probably had other problems than excessive memory or timeout (-> related to storages and handling solver prefixes)

we can test the functionality between the stage and semi-prod.
i will update some documentation and share it soon.

**Comment by VannTen** [comment]
Can do. The diff is pretty big though, so I'll push another branch with
the currents commits on work_avoidance_with_generators_splitted on my
fork, if anyone wants to have a look a that.

**Comment by VannTen** [comment]
> we can test the functionality between the stage and semi-prod. i will update some documentation and share it soon.

Any news on that ? @harshad16

**Comment by harshad16** [comment]
i have updated the documentation here for the access: https://github.com/thoth-station/thoth-application/pull/2647
with this, what we can do is run this test job , with stage and semi-prod details.

**Comment by harshad16** [comment]
will include the include the test step in this repo as well
in a different pr shortly.

**Comment by harshad16** [comment]
docs for testing https://github.com/thoth-station/document-sync-job/pull/41

**Comment by VannTen** [comment]
While testing I encountered some annoyances with packaging / having separate
files.

Would it be a problem to pull in this PR a version of #40 ? This implies to
modify the way the script is executed, but we can modify the manifest in
thoth-application right ? @harshad16

**Comment by harshad16** [comment]
> Would it be a problem to pull in this PR a version of #40 ? 

we should be able to do it. 


> This implies to modify the way the script is executed, but we can modify the manifest in thoth-application right ? @harshad16

yes we update it in  thoth-application.

**Comment by VannTen** [comment]
The date filtering does not seem to work with the older version (aka using
get_document_listing for listing objects keys) when I test it on data on upshift
("data/ocp4-test").

Makes sense since there is data not following what I can discern of the "path
schema":
```
2022-09-26 07:25:45,643 1651 DEBUG    thoth.document_sync:157: Syncing source key (bucket='thoth') 'data/ocp4-test/analysis-by-digest/sha256:4892016fc0cab7eafc8c8eabe49169ac4d8aa1e9a6985b0713e8c9cf3f582245' to dest (bucket=thoth) key '/test/pr/37/analysis-by-digest/sha256:4892016fc0cab7eafc8c8eabe49169ac4d8aa1e9a6985b0713e8c9cf3f582245'
2022-09-26 07:25:45,641 1651 DEBUG    thoth.document_sync:157: Syncing source key (bucket='thoth') 'data/ocp4-test/analysis-by-digest/sha256:2e4bbb2be6e7aff711ddc93f0b07e49c93d41e4c2ffc8ea75f804ad6fe25564e' to dest (bucket=thoth) key '/test/pr/37/analysis-by-digest/sha256:2e4bbb2be6e7aff711ddc93f0b07e49c93d41e4c2ffc8ea75f804ad6fe25564e'
2022-09-26 07:25:45,642 1651 DEBUG    thoth.document_sync:157: Syncing source key (bucket='thoth') 'data/ocp4-test/analysis-by-digest/sha256:31ccb79b1b2c2d6eff1bee0db23d5b8ab598eafd6238417d9813f1346f717c11' to dest (bucket=thoth) key '/test/pr/37/analysis-by-digest/sha256:31ccb79b1b2c2d6eff1bee0db23d5b8ab598eafd6238417d9813f1346f717c11'

...
2022-09-26 07:25:46,533 1651 DEBUG    thoth.document_sync:157: Syncing source key (bucket='thoth') 'data/ocp4-test/analysis/package-extract-210216202423-73c76b4c118c34f4' to dest (bucket=thoth) key '/test/pr/37/analysis/package-extract-210216202423-73c76b4c118c34f4'
2022-09-26 07:25:46,533 1651 DEBUG    thoth.document_sync:157: Syncing source key (bucket='thoth') 'data/ocp4-test/analysis/package-extract-210217000154-5b861347e1675df2' to dest (bucket=thoth) key '/test/pr/37/analysis/package-extract-210217000154-5b861347e1675df2'
2022-09-26 07:25:46,534 1651 DEBUG    thoth.document_sync:157: Syncing source key (bucket='thoth') 'data/ocp4-test/analysis/package-extract-210220000034-6d6173ecc5b7ca9c' to dest (bucket=thoth) key '/test/pr/37/analysis/package-extract-210220000034-6d6173ecc5b7ca9c'
```
(on the last two it should be `analysis-210220000034` as far as I can tell)
For analysis-by-digest since there is no date at all obviously it can't work.


Am I testing this wrong ?

(I checked running get_document_listing, then I get the key. But when I add the
start_date argument I get an empty result).

-----

The underlying ask is do we need the prefix filtering on dates or could we drop
that ?


@harshad16

**Comment by VannTen** [comment]
So, first benchmarks, I got that kind of result (using `time`)


Full sync (skipped are ".request" objects)
```console
2022-09-27 15:11:35,493 12266 INFO     thoth.document_sync:242: Summary of the sync operation:
2022-09-27 15:11:35,493 12266 INFO     thoth.document_sync:244: analysis-by-digest: 38 success, 0 skipped, 0 error
2022-09-27 15:11:35,494 12266 INFO     thoth.document_sync:244: analysis: 947 success, 637 skipped, 0 error
2022-09-27 15:11:35,494 12266 INFO     thoth.document_sync:244: solver: 1156 success, 0 skipped, 0 error

real	2m46.263s
user	0m26.379s
sys	0m4.593s
```

No op (everything already synced)
```console
2022-09-27 15:11:58,504 16651 INFO     thoth.document_sync:242: Summary of the sync operation:
2022-09-27 15:11:58,504 16651 INFO     thoth.document_sync:244: analysis-by-digest: 0 success, 38 skipped, 0 error
2022-09-27 15:11:58,505 16651 INFO     thoth.document_sync:244: analysis: 0 success, 1584 skipped, 0 error
2022-09-27 15:11:58,505 16651 INFO     thoth.document_sync:244: solver: 0 success, 1156 skipped, 0 error

real	0m6.791s
user	0m1.191s
sys	0m0.094s
```

I'll try to compare with the previous method later.

**Comment by harshad16** [comment]
Based on the previous question, we would be only copying the data ingestion data most of the time, like solvers, package-extract extra, so we would like to follow the prefix as `component-datetime-wfhash` like `package-extract-210220000034-6d6173ecc5b7ca9c`
we would like to work such pattern. 
thanks for all the testing and re-write this looks very promising.

**Comment by VannTen** [comment]
On Tue, Sep 27, 2022 at 12:43:20PM -0700, Harshad Reddy Nalla wrote:
> Based on the previous question, we would be only copying the data
> ingestion data most of the time, like solvers, package-extract extra,
> so we would like to follow the prefix as `component-datetime-wfhash`
> like `package-extract-210220000034-6d6173ecc5b7ca9c` we would like to
> work such pattern. 

Thing is, the prefix is not super consistent. For analysis,
analysis-by-digest, solver we have that those key:


analysis:
data/ocp4-test/analysis/package-extract-220928000012-7643569fa3a0f8b6
data/ocp4-test/analysis/package-extract-220928010100-aae17edb771ef3e
-> this is "root_prefix/adapter.RESULT_TYPE/<something>-<dateprefix>-<wfhash>"

analysis-by-digest:
data/ocp4-test/analysis/sha256:f8db086dcbf028d9a1f23f3cf9173e1de087d3d8e72a592c92c9ab510e0352f3
data/ocp4-test/analysis/sha256:fa95fe99ae271240513a91c17fd2869936286eae5c18eded1149a1005f8db383
-> this is "root_prefix/adapter.RESULT_TYPE/<hash-alg>:<hash>"

solver:
data/ocp4-test/analysis/solver-rhel-8-py36-220919010145-3f978877a8795a80
data/ocp4-test/analysis/solver-rhel-8-py36-220920010531-f033c4208b47163e
-> this is "root_prefix/adapter.RESULT_TYPE/<solver-spec>-<date-prefix>-<wfhash>"


This was not was I expected from the code in storages (and indeed,
get_document_listing does not seem to work when used with the start_date
argument, as far as I can tell, at least for analysis/analysis-by-digest)

Solver is workable, but I don't know how to go from Analysis.RESULT_TYPE
to "package-extract" for example.

> we would be only copying the data ingestion data most of the time,
If I follow correctly, this means only copying new data, right ?
The current code would do that, filtering on date is an optimization at
the start of the pipeline.

Current workflow (of this PR) is :
source = <set of keys in source S3 with correct prefix etc>
dest = <set of keys in dest S3 with correct prefix etc>
to_copy = source - dest (aka, set difference)
# so if source == dest => to_copy == []
copy_from_source_to_dest(to_copy)

Date prefixing only concerns the first step (setting source and dest)
(basically, the second case with no copying at all in my benchmarks
posted yesterday).
Comparing with empty source and destination:

$ time THOTH_DEPLOYMENT_NAME=rubbish THOTH_DOCUMENT_SYNC_DST=s3://thoth/test/pr/notapr/ document-sync-job  --dry-run 
...
2022-09-28 08:25:31,192 3373 INFO     thoth.document_sync:242: Summary of the sync operation:
2022-09-28 08:25:31,192 3373 INFO     thoth.document_sync:244: analysis-by-digest: 0 success, 0 skipped, 0 error
2022-09-28 08:25:31,192 3373 INFO     thoth.document_sync:244: analysis: 0 success, 0 skipped, 0 error
2022-09-28 08:25:31,193 3373 INFO     thoth.document_sync:244: solver: 0 success, 0 skipped, 0 error

real	0m2.829s
user	0m0.798s
sys	0m0.083s

vs

time document-sync-job --dry-run 
2022-09-28 08:25:25,609 3358 INFO     thoth.document_sync:242: Summary of the sync operation:
2022-09-28 08:25:25,609 3358 INFO     thoth.document_sync:244: analysis-by-digest: 0 success, 38 skipped, 0 error
2022-09-28 08:25:25,609 3358 INFO     thoth.document_sync:244: analysis: 2 success, 1585 skipped, 0 error
2022-09-28 08:25:25,610 3358 INFO     thoth.document_sync:244: solver: 1 success, 1156 skipped, 0 error

real	0m5.728s
user	0m1.225s
sys	0m0.058s

So we have roughly 3 seconds of overhead for 2700 keys, give or take.
That would require more testing to see how that scale though, not sure
how much is in the listing requests exactly (which are paged, 1000 keys
at a time)

**Comment by VannTen** [comment]
That's where having #2691 would be cool (date as a first class metadata, not
part of the key-name itself)

**Comment by VannTen** [comment]
Results of benchmarks on current master with commit 75079167caac966fd0d362654d524cc174584c96 on top (for the url envvar):

(the output format is different because it used time binary instead instead of
bash builtin, apparently using env parameters assignment does that)

If I remember correctly, elapsed maps to real, the rest is the same

#No-op
time python app.py --debug
...
510.59user 160.82system 3:44.24elapsed 745%CPU (0avgtext+0avgdata 145836maxresident)k
88inputs+0outputs (33major+30686525minor)pagefaults 0swaps

#Full sync
...
time python app.py --debug
3076.99user 351.99system 8:36.97elapsed 663%CPU (0avgtext+0avgdata 259352maxresident)k
0inputs+0outputs (69major+66830397minor)pagefaults 0swaps


----

If my math is correct, that's 32 times faster in the no-op case and 3 times
faster in the full-op case. (counting elapsed / real. I don't think we care
about the other much because we're mainly doing IO)

I'll see tomorrow how the dates stuff perform (if it works in master, not sure
about that).

**Comment by VannTen** [comment]
```console
[max@work-laptop-max ~/redhat/thoth-station/document_sync_job] (aws_url_endpoint =)$ time THOTH_DOCUMENT_SYNC_DST=s3://thoth/test/pr/37 python app.py --days 1 --debug
2022-09-29 09:11:20,863 11951 WARNING  thoth.common:340: Logging to a Sentry instance is turned off
2022-09-29 09:11:20,864 11951 DEBUG    thoth.document_sync:176: Debug mode is on.
2022-09-29 09:11:20,864 11951 INFO     thoth.document_sync:178: Running document syncing job in version '0.3.1+storages0.73.4'
2022-09-29 09:11:20,864 11951 INFO     thoth.document_sync:183: Listing documents created since '2022-09-28'
2022-09-29 09:11:20,864 11951 INFO     thoth.document_sync:187: Listing 'analysis-by-digest' documents
2022-09-29 09:11:21,744 11951 INFO     thoth.document_sync:187: Listing 'analysis' documents
2022-09-29 09:11:22,522 11951 INFO     thoth.document_sync:204: Listing documents for solver 'solver-rhel-8-py36'
2022-09-29 09:11:23,233 11951 INFO     thoth.document_sync:204: Listing documents for solver 'solver-rhel-8-py38'
2022-09-29 09:11:23,590 11951 INFO     thoth.document_sync:225: Document sync job has finished successfully

real	0m3.481s
user	0m0.909s
sys	0m0.070s
```
So when using `--days n` the current master version does not find any documents
it seems. (If I'm missing something please point out, this is my env:
<details>

<summary>Env</summary>

```
AWS_S3_ENDPOINT_URL=https://s3-openshift-storage.apps.smaug.na.operate-first.cloud
THOTH_CEPH_BUCKET=thoth
THOTH_CEPH_BUCKET_PREFIX=data
THOTH_CEPH_SECRET_KEY=pdcEGFERILlkRDGrCSxdIMaZVtNCOKvYP4Gf2b2x
PIPENV_ACTIVE=1
THOTH_S3_ENDPOINT_URL=https://s3.upshift.redhat.com/
THOTH_DOCUMENT_SYNC_DST=s3://thoth/test/pr/37/
EDITOR=vim
PWD=/home/max/redhat/thoth-station/document_sync_job
LOGNAME=max
SYSTEMD_EXEC_PID=815
THOTH_CEPH_KEY_ID=<xxx>
THOTH_DEPLOYMENT_NAME=ocp4-test
PIP_PYTHON_PATH=/usr/bin/python
HOME=/home/max
LANG=en_US.UTF-8
VIRTUAL_ENV=/home/max/.local/share/virtualenvs/document_sync_job-ykvVh148
AWS_SECRET_ACCESS_KEY=<xxxx>
THOTH_DOCUMENT_SYNC_CONFIGURED_SOLVERS=solver-rhel-8-py36 solver-rhel-8-py38
PIP_DISABLE_PIP_VERSION_CHECK=1
AWS_ACCESS_KEY_ID=<xxxx>
PYTHONDONTWRITEBYTECODE=1
PATH=/home/max/.local/share/virtualenvs/document_sync_job-ykvVh148/bin:/usr/local/bin:/usr/bin
```
</details>


----
Given this and the benchmarks, is it ok to drop the `--days` option for now ?

**Comment by VannTen** [comment]
This should be good for review.

The last commit removes the `--days` parameters. I put that in one commit on
its own to easily revert it if needed.

File: app/app.py [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+from thoth.storages import SolverResultsStore
+from thoth.storages import CephStore
+from thoth.storages.result_base import ResultStorageBase
+
+from prometheus_client import CollectorRegistry, Gauge, Counter as PromCounter, push_to_gateway
+from .lazy_set_ops import sorted_iter_set_difference
+
+prometheus_registry = CollectorRegistry()
+
+__version__ = "0.3.1"
```
Comment (harshad16) [inline app/app.py]: lets create a `__init__.py` and include this there 
so stay with current kebechet pattern in future we can move it in .

File: app/app.py [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+__component_version__ = f"{__version__}+storages{__thoth_storages_version__}"
+
+init_logging()
+_LOGGER = logging.getLogger("thoth.document_sync")
+
+_DEFAULT_CONCURRENCY = 16
```
Comment (harshad16) [inline app/app.py]: Lets add this as well with env vars, so we can adjust this from deployment without having to change the source code.
```suggestion
_DEFAULT_CONCURRENCY = os.getenv("THOTH_DEFAULT_CONCURRENCY",16)
```

File: app/app.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    """Sync Thoth data to a remote with an S3 compatible interface."""
+    if debug:
+        _LOGGER.setLevel(logging.DEBUG)
+        _LOGGER.debug("Debug mode is on.")
+
+    if dry_run:
+        _LOGGER.info("Running in dry-run mode")
+    _LOGGER.info("Running document syncing job in version %r", __component_version__)
+
+    src_config = CephStore(f"{os.getenv('THOTH_CEPH_BUCKET_PREFIX', '')}/{os.getenv('THOTH_DEPLOYMENT_NAME', '')}")
```
Comment (harshad16) [inline app/app.py]:
```suggestion
    src_config = CephStore(f"{os.getenv('THOTH_CEPH_BUCKET_PREFIX', '')}/{_THOTH_DEPLOYMENT_NAME}")
```

**Review by harshad16 — COMMENTED** [review]
lgtm
few minor suggestions.
can we make the directory as, `thoth/document-sync/app.py` etc 
so we can stay in sync with other project structure in thoth-station.

File: app/app.py (hunk 2 above)
Comment (VannTen) [reply app/app.py]: That's handled by the `--concurrency` flag (with click) with a var env as a default.
(I didn't added this, but seems like github is not great for file renames tracking...)

**Comment by VannTen** [comment]
Good point on using `thoth/document-sync` :+1:

File: app/app.py (hunk 1 above)
Comment (VannTen) [reply app/app.py]: I thought `__init__.py` was not necessary anymore in python modules given [PEP 420](https://peps.python.org/pep-0420/) ?

Comment (VannTen) [reply app/app.py]: (Unless you meant it's for supporting the version bump using kebechet issues ?)

Comment (harshad16) [reply app/app.py]: yeah suggestion was to stay with kebechet pattern for now, till we upgrade it.

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve

Excellent work
Thanks lets get this into action.

**Pull request merged**